internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,236 @@
+	return proxyBuilder.Build(proxyTarget, &wrappingClientConn{r, "proxy"}, opts)
+}
+
+func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
+	if r.targetResolver != nil {
+		r.targetResolver.ResolveNow(o)
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.ResolveNow(o)
+	}
+}
+
+func (r *delegatingResolver) Close() {
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.Close()
+	}
+}
+
+type keyType string
+
+const proxyConnectAddrKey = keyType("grpc.resolver.delegatingresolver.proxyConnectAddr")
+const userKey = keyType("grpc.resolver.delegatingresolver.user")
+
+// SetConnectAddr returns a copy of the provided resolver.Address with attributes
+// containing address to be sent in connect request to proxy.  It's data should
+// not be mutated after calling SetConnectAddr.
+func SetConnectAddr(resAddr resolver.Address, addr string) resolver.Address {
+	resAddr.Attributes = resAddr.Attributes.WithValue(proxyConnectAddrKey, addr)
+	return resAddr
+}
+
+// ProxyConnectAddr returns the proxy connect address in resolver.Address, or nil
+// if not present. The returned data should not be mutated.
+func ProxyConnectAddr(addr resolver.Address) string {
+	return addr.Attributes.Value(proxyConnectAddrKey).(string)
+}
+
+// SetUser returns a copy of the provided address with attributes containing
+// user info. It's data should not be mutated after calling SetUser.
+func SetUser(addr resolver.Address, user *url.Userinfo) resolver.Address {
+	addr.Attributes = addr.Attributes.WithValue(userKey, user)
+	return addr
+}
+
+// User returns the user info in the resolver.Address, or nil if not present.
+// The returned data should not be mutated.
+func User(addr resolver.Address) *url.Userinfo {
+	return addr.Attributes.Value(userKey).(*url.Userinfo)
+}
+
+func (r *delegatingResolver) updateAddressAtrributes() []resolver.Address {
+	var addresses []resolver.Address
+	for _, proxyAddr := range r.proxyAddrs {
+		for _, targetAddr := range r.targetAddrs {
+			newAddr := resolver.Address{Addr: proxyAddr.Addr}
+			newAddr = SetConnectAddr(newAddr, targetAddr.Addr)
+			if r.proxyURL.User != nil {
+				newAddr = SetUser(newAddr, r.proxyURL.User)
+			}
+			addresses = append(addresses, newAddr)
+		}
+	}
+	// return the combined addresses.
+	return addresses
+}
+
+type wrappingClientConn struct {
+	parent       *delegatingResolver
+	resolverType string
+}
+
+// UpdateState intercepts state updates from the target and proxy resolvers.
+func (wcc *wrappingClientConn) UpdateState(state resolver.State) error {
+	wcc.parent.mu.Lock()
+	defer wcc.parent.mu.Unlock()
+	var curState resolver.State
+	if wcc.resolverType == "target" {
+		wcc.parent.targetAddrs = state.Addresses
+		curState = state
+	}
+	if wcc.resolverType == "proxy" {
+		wcc.parent.proxyAddrs = state.Addresses
+	}
+
+	if len(wcc.parent.targetAddrs) == 0 || len(wcc.parent.proxyAddrs) == 0 {
+		return nil
+	}
```

Comment:
This cannot completely eat `UpdateState` calls.  If a channel has zero addresses being reported by a resolver, that is an error, and the resolver needs to receive an error in response.

Reply:
I dont see an error being returned by cc.UpdateState currently for zero addresses. As of today, the balancer returns an error for `resolver produced zero addresses` which should work for the delegating resolver as well. 

I used this condition to check if both the proxy and the target resolver have sent the update. I missed the case that the resolvers could send 0 addresses as well and it might not work correctly. 

I am changing the condition to check the updates from both resolvers and letting the balancer handle the zero address error.